src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -12,100 +12,473 @@
+// persistent stucture, because the update that is added to the work-in-progress
+// is always added to the front of the list.
+//
+// However, if incoming update is inserted into the same position of both lists,
+// we shouldn't make a copy.
+
+function insertUpdate(fiber : Fiber, update : Update, methodName : ?string) : void {
+  const queue1 = ensureUpdateQueue(fiber);
+  const queue2 = fiber.alternate ? ensureUpdateQueue(fiber.alternate) : null;
+
+  // Warn if an update is scheduled from inside an updater function.
+  if (__DEV__ && typeof methodName === 'string' && (queue1.isProcessing || (queue2 && queue2.isProcessing))) {
+    if (methodName === 'setState') {
+      console.error(
+        'setState was called from inside the updater function of another' +
+        'setState. A function passed as the first argument of setState ' +
+        'should not contain any side-effects. Return a partial state object ' +
+        'instead of calling setState again. Example: ' +
+        'this.setState(function(state) { return { count: state.count + 1 }; })'
+      );
+    } else {
+      console.error(
+        `${methodName} was called from inside the updater function of ` +
+        'setState. A function passed as the first argument of setState ' +
+        'should not contain any side-effects.'
+      );
+    }
+  }
+
+  const priorityLevel = update.priorityLevel;
+
+  let queue = queue1;
+  let insertAfter1;
+  let insertBefore1;
+  let insertAfter2;
+  let insertBefore2;
+  for (let i = 0; queue && i < 2; i++) {
+    let insertAfter = null;
+    let insertBefore = null;
+    if (queue.last && comparePriority(queue.last.priorityLevel, priorityLevel) <= 0) {
+      // Fast path for the common case where the update should be inserted at
+      // the end of the queue.
+      insertAfter = queue.last;
+    } else {
+      insertBefore = queue.first;
+      while (insertBefore && comparePriority(insertBefore.priorityLevel, priorityLevel) <= 0) {
+        insertAfter = insertBefore;
+        insertBefore = insertBefore.next;
+      }
+    }
+    if (i === 0) {
+      insertAfter1 = insertAfter;
+      insertBefore1 = insertBefore;
+      queue = queue2;
+    } else {
+      insertAfter2 = insertAfter;
+      insertBefore2 = insertBefore;
+      queue = null;
+    }
+  }
+
+  const update1 = update;
+  insertUpdateIntoQueue(queue1, update1, insertAfter1, insertBefore1);
+
+  if (queue2) {
+    let update2;
+    if (insertBefore1 === insertBefore2) {
+      // The update is inserted into the same position of both lists. There's no
+      // need to clone the update.
+      update2 = update1;
+    } else {
+      // The update is inserted into two separate positions. Make a clone of the
+      // update and insert it twice. One or the other will be dropped the next
+      // time we commit.
+      update2 = cloneUpdate(update1);
+    }
+    insertUpdateIntoQueue(queue2, update2, insertAfter2, insertBefore2);
```

Comment:
Ok I'll remove `insertUpdateIntoQueue` and inline the logic and see if that helps